lib/proxy.js

```diff
@@ -154,16 +154,32 @@ class JWProxy {
     throw new Error(`Didn't know what to do with response code ${response.statusCode}`);
   }
 
+  getSessionIdFromUrl (url) {
+    const sessionIdRe = new RegExp('/session/([^/]+)');
+    if (sessionIdRe.test(url)) {
+      const match = sessionIdRe.exec(url);
+      return match[1];
+    }
+    return false;
+  }
+
   async proxyReqRes (req, res) {
     let [response, body] = await this.proxy(req.originalUrl, req.method, req.body);
     res.headers = response.headers;
     res.set('Content-type', response.headers['content-type']);
     // if the proxied response contains a sessionId that the downstream
     // driver has generated, we don't want to return that to the client.
-    // Instead, return the id for the current session
+    // Instead, return the id from the request or from current session
     body = safeJson(body);
-    if (body && body.sessionId && this.sessionId) {
-      body.sessionId = this.sessionId;
+    if (body && body.sessionId) {
+      const reqSessionId = this.getSessionIdFromUrl(req.originalUrl);
+      if (reqSessionId) {
+        log.info(`Replacing sessionId ${body.sessionId} with ${reqSessionId}`);
+        body.sessionId = reqSessionId;
+      } else if (this.sessionId) {
+        log.info(`Replacing sessionId ${body.sessionId} with ${this.sessionId}`);
+        body.sessionId = this.sessionId;
```

Comment:
yes. so what we're doing is making sure that what we send in the json back to the client is what upstream thinks the sessionId.

i.e., we're ensuring that the sessionId never changes from the perspective of the client.